compiler/src/iree/compiler/Codegen/Common/GPU/GPUVectorDistribution.cpp

```diff
@@ -0,0 +1,263 @@
+  VectorLayoutOptions &options;
+};
+
+/// A rewriter for the pattern rewriting driver.
+class VectorDistributionRewriter : public PatternRewriter,
+                                   RewriterBase::Listener {
+public:
+  VectorDistributionRewriter(MLIRContext *ctx, DenseSet<Operation *> &erasedOps)
+      : PatternRewriter(context, this), erasedOps(erasedOps) {}
+
+  /// We need to keep track of erased operations so that we never try to
+  /// distribute them again.
+  void notifyOperationRemoved(Operation *op) override { erasedOps.insert(op); }
+
+private:
+  // Reference to operations to be erased and the worklist of the driver.
+  DenseSet<Operation *> &erasedOps;
+};
+
+static bool canDistribute(Operation *op, VectorLayoutAnalysis &analysis) {
+  auto values = llvm::to_vector_of<Value>(op->getOperands());
+  llvm::append_range(values, op->getResults());
+
+  // Check if all operands and results of this operation have a layout.
+  return llvm::all_of(values, [&](Value value) -> bool {
+    if (auto vectorValue = dyn_cast<VectorValue>(value)) {
+      return analysis.getLayout<Attribute>(vectorValue) != nullptr;
+    }
+    return false;
+  });
+}
+
+static void
+debugPrintUniqueOperationNames(SmallVectorImpl<Operation *> &worklist,
+                               DenseSet<Operation *> &exclude) {
+  DenseSet<StringRef> uniqueNames;
+  for (Operation *op : worklist) {
+    if (exclude.contains(op))
+      continue;
+    uniqueNames.insert(op->getName().getStringRef());
+  }
+
+  for (StringRef name : uniqueNames) {
+    llvm::dbgs().indent(2) << "* " << name << "\n";
+  }
+  LLVM_DEBUG(llvm::dbgs() << "\n");
+}
+
+static void applyVectorDistributionDriver(
+    Operation *root, const FrozenRewritePatternSet &patterns,
+    VectorLayoutAnalysis &analysis, VectorLayoutOptions &options) {
+
+  DenseSet<Operation *> erasedOps;
+  SmallVector<Operation *> worklist;
+
+  VectorDistributionRewriter rewriter(root->getContext(), erasedOps);
+  PatternApplicator applicator(patterns);
+  applicator.applyDefaultCostModel();
+
+  // Collect all the operations to be distributed.
+  LLVM_DEBUG(llvm::dbgs() << "Collecting operations to be distributed\n");
+  root->walk([&](Operation *op) {
+    if (canDistribute(op, analysis)) {
+      worklist.push_back(op);
+    }
+  });
+  LLVM_DEBUG(debugPrintUniqueOperationNames(worklist, erasedOps));
+
+  // Note that the pattern application here never runs on a newly created
+  // operation. It always runs on an existing operation. This ensures that no
+  // invalidated state of the analysis is ever used.
+  for (Operation *op : worklist) {
+    if (erasedOps.contains(op))
+      continue;
+
+    LLVM_DEBUG(llvm::dbgs() << "Distributing: ");
+    LLVM_DEBUG(op->print(llvm::dbgs(), OpPrintingFlags().skipRegions()));
+    LLVM_DEBUG(llvm::dbgs() << "\n");
+    if (failed(applicator.matchAndRewrite(op, rewriter))) {
+      LLVM_DEBUG(llvm::dbgs().indent(2)
+                 << ": Failed to distribute operation:\n");
```

Comment:
nit: it's a bit weird to see a trailing `:` without the op printed afterwards. I'd remove it unless it mimics some other debug print format from upstream mlir.